homeassistant/components/filesize/sensor.py

```diff
@@ -137,9 +137,8 @@ async def _async_update_data(self) -> dict[str, float | int | datetime]:
             raise UpdateFailed(f"Can not retrieve file statistics {error}") from error
 
         size = statinfo.st_size
-        last_updated = datetime.fromtimestamp(statinfo.st_mtime).replace(
-            tzinfo=dt_util.UTC
-        )
+        last_updated = dt_util.as_local(datetime.fromtimestamp(statinfo.st_mtime))
```

Comment:
I'm not following this one?

This converts a POSIX timestamp to a datetime (UTC as that is where the Epoch 0 originates from), next its converted to a local datetime, and the state engine will then convert is back to UTC? again?

Reply:
Current code replace timezone with UTC, but actually should be local. I have a offset of 2 hours in the reported sensor:

ls -la:
`-rw-r--r-- 1 root root         5 Apr 15 20:07 test_time`

sensor:
<img width="622" alt="image" src="https://user-images.githubusercontent.com/57354320/163691533-b1b047a4-f658-445e-8d5c-4d6a24b19d28.png">

my fix:
![image](https://user-images.githubusercontent.com/57354320/163691543-4f1825b4-6855-4572-8d5a-85d85cb018a5.png)

Simone

Reply:
> Current code replace timezone with UTC, but actually should be local. I

That is the problem, that makes no sense, as a POSIX timestamp is always from UTC...

Reply:
Sensor  reads `1650053220.0` as POSIX timestamp (`statinfo.st_mtime`).
It converts to **Friday, April 15, 2022 8:07:00 PM**

If you consider this value UTC, then it's wrong. As 20.07 is local time as shown in the ls output above.

Simone

Reply:
>  As 20.07 is local time as shown in the ls output above.

Maybe that is a display thing?

In any case, POSIX time is not localized. It is the number of seconds that have elapsed since the Unix epoch. The Unix epoch is 00:00:00 UTC on 1 January 1970.

Handling POSIX timestamps as non-UTC can't be correct.

Reply:
> > As 20.07 is local time as shown in the ls output above.
> 
> Maybe that is a display thing?

Don't this so, the fact is that the file was last changed by me at 20:07 local time, not at 20:07 UTC.

> 
> In any case, POSIX time is not localized. It is the number of seconds that have elapsed since the Unix epoch. The Unix epoch is 00:00:00 UTC on 1 January 1970.
> 
> Handling POSIX timestamps as non-UTC can't be correct.

I think that this is a bug(feature?) on how `os.stat` return values.

Simone

Reply:
`datetime.fromtimestamp` returns the local time (https://github.com/python/cpython/blob/3.10/Lib/datetime.py#L829)

By replacing the timezone with UTC the time is wrong because its replacing the local timezone with UTC without converting it

Reply:
Suggested fix https://github.com/home-assistant/core/pull/70114/files#r851726762

Reply:
> Suggested fix https://github.com/home-assistant/core/pull/70114/files#r851726762

Already tested, but it reports 20:07

Reply:
The issue here is that the raw value reported by `os.stat` is local time and not UTC.

Reply:
Running `stat -c '%y' config/test_time`  from command line shows the issue as well:

2022-04-15 20:07:00.000000000 +0000

Simone

Reply:
The `stat` binary converts the timestamp from the system call to local time.

In the example below, the host is set to HST, and in the docker container the timezone is set to UTC.
```
root@mauinasa:~# stat -c '%y' /volume1/docker/89-homeassistant/configuration.yaml 
2022-04-16 21:32:19.195471302 -1000
root@mauinasa:~#  docker exec -it 89-homeassistant /bin/bash
bash-5.1# stat -c '%y' /config/configuration.yaml 
2022-04-17 07:32:19.195471302 +0000
```